src/scala/microsoft-spark-2.4.x/src/main/scala/org/apache/spark/api/dotnet/DotnetBackend.scala

```diff
@@ -81,30 +79,17 @@ class DotnetBackend extends Logging {
     bootstrap = null
 
     // Send close to .NET callback server.
-    logInfo("Requesting to close all call back sockets")
-    var socket: Socket = null
-    do {
-      socket = DotnetBackend.callbackSockets.poll()
-      if (socket != null) {
-        try {
-          val dos = new DataOutputStream(socket.getOutputStream)
-          SerDe.writeString(dos, "close")
-          socket.close()
-          socket = null
-        } catch {
-          case e: Exception => logError("Exception when closing socket: ", e)
-        }
-      }
-    } while (socket != null)
-    DotnetBackend.callbackSocketShutdown = true
+    DotnetBackend.shutdownCallbackClient()
   }
 }
 
 object DotnetBackend {
-  // Channels to callback server.
-  private[spark] val callbackSockets: BlockingQueue[Socket] = new LinkedBlockingQueue[Socket]()
-  @volatile private[spark] var callbackPort: Int = 0
+  @volatile private[spark] var callbackClient: CallbackClient = null
 
-  // flag to denote whether the callback socket is shutdown explicitly
-  @volatile private[spark] var callbackSocketShutdown: Boolean = false
+  private[spark] def shutdownCallbackClient(): Unit = {
```

Comment:
should this be `synchronized`?

Reply:
If it does, then we should make setting synchronized.

Reply:
yes good catch